RefApp/ViewControllers/PlayerViewController.swift

```diff
@@ -0,0 +1,448 @@
+                guard let `self` = self else { return }
+                self.togglePlayPauseButton(paused: false)
+            }
+            .onPlaybackPaused{ [weak self] player, source in
+                // Fires when the playback pauses for some reason
+                guard let `self` = self else { return }
+                self.togglePlayPauseButton(paused: true)
+            }
+            .onPlaybackResumed{ [weak self] player, source in
+                // Fires when the playback resumes from a paused state
+                guard let `self` = self else { return }
+                self.togglePlayPauseButton(paused: false)
+            }
+            .onPlaybackAborted{ player, source in
+                // Published once the player.stop() method is called.
+                // This is considered a user action
+            }
+            .onPlaybackCompleted{ player, source in
+                // Published when playback reached the end of the current media.
+        }
+        
+        // Besides playback control events Player also publishes several status related events.
+        player
+            .onProgramChanged { [weak self] player, source, program in
+                // Update user facing program information
+                guard let `self` = self else { return }
+                self.update(withProgram: program)
+            }
+            .onEntitlementResponse { [weak self] player, source, entitlement in
+                // Fires when a new entitlement is received, such as after attempting to start playback
+                guard let `self` = self else { return }
+                self.update(contractRestrictions: entitlement)
+            }
+            .onBitrateChanged{ player, source, bitrate in
+                // Published whenever the current bitrate changes
+                //self?.updateQualityIndicator(with: bitrate)
+            }
+            .onBufferingStarted{ player, source in
+                // Fires whenever the buffer is unable to keep up with playback
+            }
+            .onBufferingStopped{ player, source in
+                // Fires when buffering is no longer needed
+            }
+            .onDurationChanged{ player, source in
+                // Published when the active media received an update to its duration property
+        }
+        
+        // Error handling can be done by listening to associated event.
+        player
+            .onError{ [weak self] player, source, error in
+                guard let `self` = self else { return }
+                let okAction = UIAlertAction(title: NSLocalizedString("Ok", comment: ""), style: .cancel, handler: {
+                    (alert: UIAlertAction!) -> Void in
+                })
+                
+                let message = "\(error.code) " + error.message + "\n" + (error.info ?? "")
+                self.popupAlert(title: error.domain , message: message, actions: [okAction], preferedStyle: .alert)
+        }
+        
```

Comment:
I think we should add the `onWarning` callback too, it is useful for demonstrating contract restrictions

Reply:
Ok, will add the `onWarning` callback too